### PR TITLE
Use local kubectl for k8s deployment

### PR DESCRIPTION
**Pull request by atsikham**
#1615, #1616 and #1659 have to be implemented before merge
Verified by manual execution of @seriva [poc](https://github.com/epiphany-platform/epiphany/issues/1616#issuecomment-691019357) before applying configuration.

It's supposed that `kubeconfig` is located under `inventory_dir`. KUBECONFIG is specified on the play level and defined in `group_vars`.

**Untouched:**

- applications - separate task #1654 
- vault - separate task #1672 
- usage of `kubectl` with `kubelet.conf` as `kubelet.conf` is located only on remote hosts and tasks relates to these hosts:
:small_orange_diamond: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_common/tasks/gather-facts.yml
:small_orange_diamond: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/verify-upgrade.yml
- bash completion setup on remote host
:small_orange_diamond: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/tasks/main.yml
- preflight
:small_orange_diamond: core/src/epicli/data/common/ansible/playbooks/roles/preflight_facts/tasks/kubernetes.yml. Explanation is [here](https://github.com/epiphany-platform/epiphany/pull/1663#discussion_r490091462)

**Verified locally with Ubuntu+AWS configuration:**

- apply:
:white_check_mark: single control plane node
:white_check_mark: promotion to HA
:white_check_mark: HA k8s cluster
:white_check_mark: k8s cluster with worker nodes
:white_check_mark: k8s + monitoring (prometheus) and rabbitmq
:white_check_mark: certificates renewal
- upgrade:
:white_check_mark: k8s with included #1659 fix
:white_check_mark: k8s + monitoring (prometheus) and rabbitmq
- backup
:x: backup with following configuration 

```yaml
---
kind: configuration/backup
title: Backup Config
name: default
specification:
  components:
    load_balancer:
      enabled: false
    logging:
      enabled: false
    monitoring:
      enabled: true
    postgresql:
      enabled: false
    rabbitmq:
      enabled: true
    kubernetes:
      enabled: true
```

Backup does not seem to work correctly itself. With HA cluster got such an error, at least etcd image name is not processed to have value only from one pod:

```text
16:43:33 INFO cli.engine.ansible.AnsibleCommand - TASK [backup : Save etcd snapshot] ******************************************************************************************************
16:43:35 ERROR cli.engine.ansible.AnsibleCommand - fatal: [ec2-3-95-156-60.compute-1.amazonaws.com]: FAILED! => {"changed": true, "cmd": "docker run -v \"/epibackup/ansible.zKwY0W.tmp/:/backup/\" --network host --env ETCDCTL_API=3 --rm \"ec2-3-95-156-60.compute-1.amazonaws.com:5000/k8s.gcr.io/etcd:3.4.3-0 ec2-3-95-156-60.compute-1.amazonaws.com:5000/k8s.gcr.io/etcd:3.4.3-0 ec2-3-95-156-60.compute-1.amazonaws.com:5000/k8s.gcr.io/etcd:3.4.3-0\" etcdctl --endpoints https://127.0.0.1:2379 --cacert /backup/pki/etcd/ca.crt --cert /backup/pki/etcd/healthcheck-client.crt --key /backup/pki/etcd/healthcheck-client.key snapshot save /backup/etcd-snapshot.db\n", "delta": "0:00:00.062874", "end": "2020-09-16 16:43:37.864740", "msg": "non-zero return code", "rc": 125, "start": "2020-09-16 16:43:37.801866", "stderr": "docker: invalid reference format.\nSee 'docker run --help'.", "stderr_lines": ["docker: invalid reference format.", "See 'docker run --help'."], "stdout": "", "stdout_lines": []}
```

Created new issue - #1664.

**Timings:**

Apply configuration for k8s cluster (3 control plane + 2 worker nodes), prometheus and rabbitmq with control plane certificates renewal:

With changes - 2572032ms
Without changes - 2523567ms

At the first glance deployment time is not worse. It depends on different affecting factors, so it's not enough to use such amount of results for detailed comparison.

### CODE REVIEW

File: core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml [hunk 1]
```diff
@@ -22,16 +22,16 @@
...
+  shell: "kubectl apply -f /home/{{ admin_user.name }}/k8s-rolebinding.yml"
   run_once: true
-  delegate_to: "{{ master_hostname }}"
+  delegate_to: localhost
 
 - name: Get kubernetes bearer token for prometheus
   become: yes
   become_user: "{{ admin_user.name }}"
-  shell: "kubectl --kubeconfig=/home/{{ admin_user.name }}/.kube/config describe secret $(kubectl get secrets --namespace=kube-system | grep prometheus | awk '{print $1}') --namespace=kube-system | grep -E '^token' | awk '{print $2}' | head -1"
+  shell: "kubectl describe secret $(kubectl get secrets --namespace=kube-system | grep prometheus | awk '{print $1}') --namespace=kube-system | grep -E '^token' | awk '{print $2}' | head -1"
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml:32]: I don't like this line, too much bash magic I think we should refactor it. 🤔

**Review by sk4zuzu — COMMENTED** [review]
Very nice cleanup. 👍 I'll look more into it.

File: core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml (hunk 1 above)
Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml:32]: How about this one?

kubectl get $(kubectl get secrets -n kube-system -o name | grep prometheus) -n kube-system -o jsonpath='{ .data.token }'

Comment (sk4zuzu) [reply core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml:32]: Haven't tested it, but looks much better! :+1:

**Comment by sk4zuzu** [comment]
@atsikham I'm noticing `delegate_to: localhost` statements in `kubernetes_master` and `kubernetes_node` roles as well as in `upgrade / kuberenetes`. I don't think we need those changes. This code will never run on AKS or EKS. WDYT? 🤔

**Comment by atsikham** [comment]
> @atsikham I'm noticing `delegate_to: localhost` statements in `kubernetes_master` and `kubernetes_node` roles as well as in `upgrade / kuberenetes`. I don't think we need those changes. This code will never run on AKS or EKS. WDYT? 🤔

@sk4zuzu there are some improvements, such as running tasks only once, store files locally and do not copy them to the remote host. Also kubeconfig var is defined only in one place and avoided switching to bash where it's possible. Agree that we can leave it as is, but I'll finish this work soon and as it hasn't been merged yet, we can decide.

File: core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml (hunk 1 above)
Comment (atsikham) [reply core/src/epicli/data/common/ansible/playbooks/roles/prometheus/tasks/configure-k8s-apps-monitoring.yml:32]: image

**Comment by atsikham** [comment]
@seriva @sk4zuzu @to-bar @mkyc could you follow this discussion. We need to decide if these changes will be merged.

File: core/src/epicli/cli/epicli.py [hunk 2]
```diff
@@ -217,12 +217,15 @@ def upgrade_parser(subparsers):
...
                             help='Path to the folder with pre-prepared offline requirements.')
+    sub_parser.add_argument('--vault-password', dest='vault_password', type=str,
+                            help='Password that will be used to encrypt build artifacts.')
     # developer options
     sub_parser.add_argument('--profile-ansible-tasks', dest='profile_ansible_tasks', action="store_true",
                             help='Enable Ansible profile_tasks plugin for timing tasks.')
 
     def run_upgrade(args):
         adjust_paths_from_build(args)
+        ensure_vault_password_is_set(args)
```
Comment (atsikham) [inline core/src/epicli/cli/epicli.py:228]: This is done to be able to use `copy-kubeconfig` tasks in `upgrade` playbook

**Comment by sk4zuzu** [comment]
@atsikham I think we should have `become: false` in every place where we run `delegate_to: localhost` :thinking: Otherwise if anybody runs epicli on a local account without sudo privileges elevated, it will fail. :(

**Comment by sk4zuzu** [comment]
On the current HEAD I get:
```
08:45:18 INFO cli.engine.ansible.AnsibleCommand - TASK [kubernetes_master : Apply /workspaces/epiphany/core/src/epicli/clusters/build/any1/manifests/users.yml file] **************************
08:45:18 INFO cli.engine.ansible.AnsibleCommand - task path: /workspaces/epiphany/core/src/epicli/clusters/build/any1/ansible/roles/kubernetes_master/tasks/deployments/deploy-file.yml:22
08:45:18 ERROR cli.engine.ansible.AnsibleCommand - fatal: [x1a1 -> localhost]: FAILED! => {"changed": false, "cmd": ["kubectl", "apply", "-f", "/workspaces/epiphany/core/src/epicli/clusters/build/any1/manifests/users.yml"], "delta": "0:00:00.336727", "end": "2020-09-22 08:45:18.779222", "msg": "non-zero return code", "rc": 1, "start": "2020-09-22 08:45:18.442495", "stderr": "W0922 08:45:18.771820   14775 loader.go:223] Config not found: /workspaces/epiphany/core/src/epicli/clusters/build/any1/kubeconfig\nThe connection to the server localhost:8080 was refused - did you specify the right host or port?", "stderr_lines": ["W0922 08:45:18.771820   14775 loader.go:223] Config not found: /workspaces/epiphany/core/src/epicli/clusters/build/any1/kubeconfig", "The connection to the server localhost:8080 was refused - did you specify the right host or port?"], "stdout": "", "stdout_lines": []}
```
I don't have `kubeconfig` file inside my build dir. :thinking:

**Comment by sk4zuzu** [comment]
Moreover if I don't provide the `kubeconfig` myself it fails every time I re-run `epicli apply`. :thinking:

File: core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/defaults/main.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ cni_plugin_vars:
     canal: k8s-app=canal
     flannel: app=flannel
 
-epiphany_manifests_dir: /etc/epiphany/manifests
+epiphany_manifests_dir: "{{ inventory_dir }}/manifests"
```
Comment (to-bar) [inline core/src/epicli/data/common/ansible/playbooks/roles/kubernetes_master/defaults/main.yml:8]:
```suggestion
epiphany_k8s_manifests_dir: "{{ inventory_dir }}/k8s_manifests"
```
After changing variable name, we will need to adjust all references.

**Review by to-bar — CHANGES_REQUESTED** [review]

**Comment by atsikham** [comment]
Closed as there is another PR #1690 with necessary changes exists.

**Pull request closed without merging**